ローカルの CPP SDK を参照してビルドするためのオプションを追加する

enm10k opened the pull request:
## 変更内容

- 標題の機能を実現するため、ビルド・スクリプトに以下のオプションを追加しました
  - `--sora-dir` ... ローカルの Sora C++ SDK のパスを絶対パスで指定します。 Sora C++ SDK はビルドされます。
  - `--sora-args` ... ローカルの Sora C++ SDK をビルドする際のオプションを指定します

## 残タスク

sumomo の macos_arm64, windows_x86_64, ubuntu-20.04_armv8_jetson のみ対応しています。
方針に問題が無いかレビューを受けた後、他のサンプル/プラットフォームの対応 & ドキュメンテーションを行います。

## 現状把握している課題

`--sora-args` でオプションを指定する際に `=` を使用しないと正しく認識されません

```
# OK
$ python3 sumomo/macos_arm64/run.py --sora-dir $HOME/github.com/shiguredo/hoge/sora-cpp-sdk --sora-args='--test'

# NG
$ python3 sumomo/macos_arm64/run.py --sora-dir $HOME/github.com/shiguredo/hoge/sora-cpp-sdk --sora-args '--test'
usage: run.py [-h] [--debug] [--sora-dir SORA_DIR] [--sora-args SORA_ARGS]
run.py: error: argument --sora-args: expected one argument
```

---

This pull request primarily focuses on refactoring the way dependencies are handled in the build scripts for the `messaging_recvonly_sample` project across different platforms. The changes aim to improve code reusability and maintainability by introducing a new function `install_sora_and_deps`, which installs the Sora C++ SDK, Boost, and Lyra. If a directory for Sora is provided, the function `build_sora` is used instead. The new `SoraInfo` class is introduced to store information about the installation directories of these dependencies.

The most important changes are:

**Changes to `base.py`:**

* Imported the `shlex` module.
* Added a new class `SoraInfo` to hold the installation directories of Sora, Boost, and Lyra.
* Added new functions `install_sora_and_deps`, `add_sora_arguments`, `build_sora`, and `get_sora_info` to handle the installation of Sora and its dependencies.

**Changes to `messaging_recvonly_sample` scripts:**

* Imported `Optional` and `List` from the `typing` module and the new functions from `base.py` in all the `run.py` scripts. [[1]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4R5) [[2]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fR6) [[3]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182R5) [[4]](diffhunk://#diff-167d1578e7e162177a0ed7088e172e700607a5302d6ba4081468befd2ff23326R5)
* Refactored the `install_deps` function in all the `run.py` scripts to use `install_sora_and_deps` or `build_sora` instead of installing Boost, Lyra, and Sora separately. [[1]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4R19-R30) [[2]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fR19-R32) [[3]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182R18-R31) [[4]](diffhunk://#diff-167d1578e7e162177a0ed7088e172e700607a5302d6ba4081468befd2ff23326R18-R30)
* Added `add_sora_arguments` to the argument parser in the `main` function of all the `run.py` scripts. [[1]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4R74) [[2]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fR114) [[3]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182R100) F6baab7bL133R110)
* Modified the `main` function in all the `run.py` scripts to use `get_sora_info` and pass the installation directories to `cmake_args`. [[1]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4L110-R103) [[2]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fL150-R143) [[3]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182L136-R129) F6baab7bL147R134)

melpon commented on sumomo/macos_arm64/run.py:
```diff
@@ -140,13 +116,18 @@ def main():
+        if args.sora_dir:
+            sora_info = get_sora_info(os.path.join(args.sora_dir, '_install', dir, configuration_dir))
+        else:
+            sora_info = get_sora_info(install_dir)
```
sora_dir, install_dir, dir, debug を引数に渡せば run.py では分岐がいらなくなるはず。

melpon commented on base.py:
```diff
@@ -395,6 +395,70 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+    parser.add_argument("--sora-args", type=str, default="")
+
+
+def build_sora(platform: str, sora_dir: str, sora_args: str, debug: bool):
```
- `sora_args` が str 型なのが微妙なので、事前にリストにしておいた方が良さそう
- リストにするのは `argparse` の `type=` で関数を渡してやるのが綺麗になりそう
- `type=` に `str.split()` を渡してもいいけど、[`shlex.split()`](https://docs.python.org/ja/3/library/shlex.html#shlex.split) を渡した方がエスケープ処理とかもうまいことやってくれるのでこっちの方が良さそう

melpon commented on base.py:
```diff
@@ -395,6 +395,70 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+def build_sora(platform: str, sora_dir: str, sora_args: str, debug: bool):
+    args = sora_args.split()
+    if debug and '--debug' not in args:
+        args.insert(0, '--debug')
```
`args = ['--debug', *args]` の方が分かりやすいです。

melpon commented on base.py:
```diff
@@ -395,6 +395,70 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+
+
+def add_sora_arguments(parser):
+    parser.add_argument("--sora-dir", type=str, default="")
```
`default=""` は指定しないで下さい。未指定なら `None` になってた方が分かりやすいです。

enm10k commented:
@melpon 
これまでいただいた以下の指摘について、 [6987617](https://github.com/shiguredo/sora-cpp-sdk-samples/pull/44/commits/6987617ffd4498ed73784ac810987ee5de188df0) で対応しました

- get_sora_info 内で条件をハンドルし、呼び出し側の分岐を減らす
- `--sora-dir` のデフォルト引数を None に変更 (それに伴い type hint を Optional[str] に変更)
- `--sora-args` の type に shlex.split を渡す & sora_args の type hint を List[str] に変更
- `args.insert(0, '--debug')` をやめて `args = ['--debug', *args]` する

melpon commented on sumomo/macos_arm64/run.py:
```diff
@@ -42,27 +44,11 @@ def install_deps(source_dir, build_dir, install_dir, debug):
-        }
-        install_lyra(**install_lyra_args)
+        # Sora C++ SDK, Boost, Lyra
+        if sora_dir:
```
Optional 型は `is None` か `is not None`で比較してください。
else 節があるから、ここでは `is None` にして処理を入れ替えるのがいいかな

melpon commented on base.py:
```diff
@@ -395,6 +396,72 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+
+
+def get_sora_info(install_dir: str, sora_dir: Optional[str], platform: str, configuration: str) -> SoraInfo:
+    if sora_dir:
```
Optional 型は `is None` か `is not None`で比較してください。

melpon commented on base.py:
```diff
@@ -395,6 +396,72 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+        cmd(['python3', 'run.py', platform, *sora_args])
+
+
+def get_sora_info(install_dir: str, sora_dir: Optional[str], platform: str, configuration: str) -> SoraInfo:
```
ここでは configuration じゃなくて debug を受け取った方がいいです。
sora_dir が指定された時のインストール先ディレクトリの構築はこの関数の役目になるから、configuration がどうやって構築されるか呼び出し元は知らなくて良いはず。

enm10k commented:
@melpon 
新たにいただいた3点の指摘について [6fbd13b](https://github.com/shiguredo/sora-cpp-sdk-samples/pull/44/commits/6fbd13b91a076489988cf1839823f7d011c7b1ee) で修正しました

melpon commented:
良さそう。これで全部に適用しちゃってください。

melpon commented:
ただ、VSCode の Python 設定の仕様が変わって多分今は linter が動作してないと思うので https://github.com/shiguredo/sora-cpp-sdk/commit/4b9538639b2275666046624b7784a88e89960ffb#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357 あたりを参考に直しておいて下さい

enm10k commented:
@melpon 
https://github.com/shiguredo/sora-cpp-sdk-samples/pull/44#issuecomment-1969660727

こちらは Sora C++ SDK を参考に ruff や mypy を設定することになったので、別の PR で対応しようと思います
(作業ボリュームが増えたためです)

この PR で対応すべきであれば再度指示してください

enm10k commented:
@melpon
全てのビルド・スクリプトに修正を適用しました
ご確認お願いいたします

また、この PR で追加したオプションの説明を書きました
内容は以下のとおりです

```
$ python3 sumomo/macos_arm64/run.py --help
usage: run.py [-h] [--debug] [--sora-dir SORA_DIR] [--sora-args SORA_ARGS]

options:
  -h, --help            show this help message and exit
  --debug
  --sora-dir SORA_DIR   Refer to local Sora C++ SDK. When this option is specified, Sora C++ SDK will also be built. Specifying an absolute path is
                        recommended.
  --sora-args SORA_ARGS
                        Options for building local Sora C++ SDK when `--sora-dir` is specified.
```

---

This pull request includes changes to the `base.py` and `run.py` files across different platforms to simplify the installation of dependencies and improve the codebase. The main changes involve the introduction of a new class `SoraInfo` and functions `install_sora_and_deps`, `add_sora_arguments`, `build_sora`, and `get_sora_info`. These changes aim to streamline the installation process of the Sora C++ SDK, Boost, and Lyra, and provide flexibility in specifying the Sora C++ SDK directory and arguments.

Changes in `base.py`:

* [`base.py`](diffhunk://#diff-8a3053a35fc794b288420f7343bfbc575c18bc834c223c2aae201e0adcfdd5c0R7): Added `shlex` to the list of imports to support the splitting of Sora C++ SDK build arguments.
* [`base.py`](diffhunk://#diff-8a3053a35fc794b288420f7343bfbc575c18bc834c223c2aae201e0adcfdd5c0R399-R469): Introduced the `SoraInfo` class to store the installation directories of Sora C++ SDK, Boost, and Lyra. Also added functions `install_sora_and_deps`, `add_sora_arguments`, `build_sora`, and `get_sora_info` to handle the installation of Sora C++ SDK and its dependencies.

Changes in `run.py` across different platforms:

* `messaging_recvonly_sample/macos_arm64/run.py`, `messaging_recvonly_sample/ubuntu-20.04_armv8_jetson/run.py`, `messaging_recvonly_sample/ubuntu-20.04_x86_64/run.py`, `messaging_recvonly_sample/ubuntu-22.04_x86_64/run.py`: Imported `get_sora_info`, `build_sora`, `install_sora_and_deps`, `add_sora_arguments` from `base.py`, and replaced individual installations of Boost, Lyra, and Sora C++ SDK with a call to `install_sora_and_deps` or `build_sora` depending on the presence of a Sora directory. Also, added `sora_dir` and `sora_args` as arguments to `install_deps`. [[1]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4R19-R30) [[2]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4L42-R48) [[3]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4L76-L85) [[4]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4R74) [[5]](diffhunk://#diff-875562cbde91176f1a2345c2b5af4da9c5c01402bb265cf94065b9b1af94f7d4L110-R103) [[6]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fR19-R32) [[7]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fL82-R88) [[8]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fL116-L125) [[9]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fR114) [[10]](diffhunk://#diff-362f7c86fb12c4463e2ed6ae063339b27f681d55cf3cbb3b6c93ff1690f23c9fL150-R143) [[11]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182R18-R31) [[12]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182L68-R74) [[13]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182L102-L111) [[14]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182R100) [[15]](diffhunk://#diff-71be74e9400c9082715466e62ebb5a4e79101dd5bcbceff935718f06184dc182L136-R129) [[16]](diffhunk://#diff-167d1578e7e162177a0ed7088e172e700607a5302d6ba4081468befd2ff23326R18-R30)

These changes provide a more flexible and streamlined process for installing the Sora C++ SDK and its dependencies across different platforms.

melpon commented on base.py:
```diff
@@ -395,6 +396,77 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+    parser.add_argument("--sora-dir", type=str, default=None,
+                        help="Refer to local Sora C++ SDK. "
+                        "When this option is specified, Sora C++ SDK will also be built. "
+                        "Specifying an absolute path is recommended.")
```
相対パスでも問題なく動いてほしいです。今後多分 examples の下に行った時に python3 sumomo/macos_arm64/run.py --sora-dir ../ とか、python3 run.py --sora-dir ../../../ とかやるので。

enm10k replied:
@melpon [3c0f000](https://github.com/shiguredo/sora-cpp-sdk-samples/pull/44/commits/3c0f000cb5a8ab5df797b17c12f1752c20535e76) で対応しました

以下のパターンで動作を確認しました

```
# --sora-dir を絶対パスで指定
python3 sumomo/macos_arm64/run.py --sora-dir $HOME/github.com/shiguredo/hoge/sora-cpp-sdk --sora-args='--test'

# --sora-dir を相対パスで指定 その1
python3 sumomo/macos_arm64/run.py --sora-dir ../sora-cpp-sdk --sora-args='--test'

# --sora-dir を相対パスで指定 その2
cd sumomo/macos_arm64 && python3 run.py --sora-dir ../../../sora-cpp-sdk --sora-args='--test'
```

melpon commented on base.py:
```diff
@@ -395,6 +396,80 @@ def get_webrtc_info(webrtcbuild: bool, source_dir: str, build_dir: str, install_
+    install_sora(**install_sora_args)
+
+
+'''
```
docstring で書くなら docstring の作法に従って下さい

melpon replied:
ついでに —sora-args の注意点も書いておくといいかも（=で書かないといけないやつ）

enm10k replied:
[3c6412d](https://github.com/shiguredo/sora-cpp-sdk-samples/pull/44/commits/3c6412d9078f67995cd9d9637cab511d97f2b26c) で修正しました

引数の説明するのが主目的ではないため、 (docstring ではなく) 普通のコメントで add_sora_arguments の説明を残しました

melpon reviewed (approved):
よさそう

enm10k commented:
レビューありがとうございます
マージします
